server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3263,73 +3273,89 @@ SnapshotsInProgress computeUpdatedState() {
         }
 
-        void completeWithUpdatedState(SnapshotsInProgress snapshotsInProgress) {
-            if (updatesByRepo.isEmpty() == false) {
-                final var result = new ShardSnapshotUpdateResult(initialState.metadata(), snapshotsInProgress);
-                try (
-                    var onCompletionRefs = new RefCountingRunnable(
-                        () -> completionHandler.handleCompletion(result, newlyCompletedEntries, updatesByRepo.keySet())
-                    )
-                ) {
-                    for (final var taskContext : batchExecutionContext.taskContexts()) {
-                        if (taskContext.getTask() instanceof ShardSnapshotUpdate task) {
-                            final var ref = onCompletionRefs.acquire();
-                            taskContext.success(() -> {
-                                try (ref) {
-                                    task.listener.onResponse(result);
-                                }
-                            });
-                        }
+        /**
+         * Reply to all the listeners of the {@link ShardSnapshotUpdate} tasks in {@link #batchExecutionContext}; then run the final
+         * callback {@link #completionHandler}.
+         */
+        void completeCallbacksWithUpdatedState(SnapshotsInProgress snapshotsInProgress) {
+            if (updatesByRepo.isEmpty()) {
+                return;
+            }
+
+            final var result = new ShardSnapshotUpdateResult(initialState.metadata(), snapshotsInProgress);
+            try (
+                // Run the post updates callback (handleShardSnapshotUpdateCompletion) after replying to the individual update requests.
+                var onCompletionRefs = new RefCountingRunnable(
+                    () -> completionHandler.handleCompletion(result, newlyCompletedEntries, updatesByRepo.keySet())
+                )
+            ) {
+                for (final var taskContext : batchExecutionContext.taskContexts()) {
+                    if (taskContext.getTask() instanceof ShardSnapshotUpdate task) {
+                        final var ref = onCompletionRefs.acquire();
+                        taskContext.success(() -> {
+                            try (ref) {
+                                task.listener.onResponse(result);
+                            }
+                        });
                     }
                 }
             }
         }
 
-        private SnapshotsInProgress.Entry applyToEntry(SnapshotsInProgress.Entry entry, List<ShardSnapshotUpdate> updates) {
+        /**
+         * Iterates the shard snapshot updates for those applicable to the particular snapshot entry, and applies the updates.
+         * <p>
+         * NB: this method may remove entries from the underlying data structure to which {@code shardUpdates} refers.
+         */
+        private SnapshotsInProgress.Entry applyUpdatesToEntry(SnapshotsInProgress.Entry entry, Iterator<ShardSnapshotUpdate> shardUpdates) {
```

Comment:
I don't like the change from `List` to `Iterator` here. It's important that each entry gets a fresh `Iterator` over the underlying list, but by pulling this up to the caller we lose that constraint. I mean it is still correct even with this change, but it's a little subtle and if you are convinced we need it then I'd rather it was its own PR rather than hidden in this one that's otherwise just renames & comments.